### PR TITLE
Agent add support for multiples outputs

### PR DESCRIPTION
**Pull request by ph**
This PR takes a single configuration with multiple outputs configured and will create unique configuration per output.

Multiple output example
```yaml
outputs:
  default:
    type: es
    url: xxxxx
    username: xxx
    pass:  xxx
    index_name: general
  infosec1:
    type: es
    url: xxxxx
    username: xxx
    pass:  xxx
    spool:
        file: "${path.data}/spool.dat"

streams:
  - type: log
    path: /xxxx
    processors:
      - dissect:
          tokenizer: "---"
    output:
      # use_output: default
      override:
        index_name: my_service_logs
        ingest_pipeline: process_logs

  - type: log
    path: /xxxx
    output:
      # use_output: default
      index_name: mysql_access_logs

  - type: metrics/system
    username: xxxx
    pass: yyy
    output:
      use_output: infosec1
      index_name: mysql_metrics
```

fixes: #14445 

TODO: 
- [ ] update the configuration YAML for streams.

### CODE REVIEW

**Comment by ph** [comment]
@michalpristas Is still in draft but if you can an early look.

**Comment by ph** [comment]
Jenkins test this please

**Comment by michalpristas** [comment]
could you add some sample transformation into description?

File: x-pack/agent/pkg/agent/program/program.go [hunk 1]
```diff
@@ -90,3 +111,146 @@ func KnownProgramNames() []string {
 	}
 	return names
 }
+
+func groupBy(single *transpiler.AST) (map[string]*transpiler.AST, error) {
```
Comment (michalpristas) [inline x-pack/agent/pkg/agent/program/program.go]: minor thing, can you rename it to groupByOutput or something similar. i would expect additional argument describing grouping with `groupBy`

File: x-pack/agent/pkg/agent/program/program.go [hunk 2]
```diff
@@ -90,3 +111,146 @@ func KnownProgramNames() []string {
...
+	// Recreates multiple configuration grouped by the name of the outputs.
+	// Each configuration will be started into his own operator with the same name as the output.
+	grouped := make(map[string]map[string]interface{})
+
+	m, ok := normMap[outputsKey]
+	if !ok {
+		return nil, errors.New("fail to received a list of configured outputs")
+	}
+
+	out := m.(map[string]interface{})
```
Comment (michalpristas) [inline x-pack/agent/pkg/agent/program/program.go]: this is provided by user we should check conversion is correct

File: x-pack/agent/pkg/agent/program/program.go (hunk 1 above)
Comment (ph) [reply x-pack/agent/pkg/agent/program/program.go]: Make sense, I think I was forward thinking too much here.

**Comment by ph** [comment]
@michalpristas I believe I have addresses all your concerns in the review and I also rebased the PR with the fleet branch.

**Comment by ph** [comment]
I am going to update the configuration in another PR.

**Comment by ppf2** [comment]
Can you clarify if this will allow users to configure (within a single beat) to send monitoring metrics to a remote ES monitoring cluster (output.elasticsearch) and non-monitoring metrics (actual event data) to a different output (e.g., output.kafka).   Or is this use case not covered by this PR and is a different enhancement?  Thx!

**Comment by ph** [comment]
@ppf2 This doesn't change the behavior of beats, "single beat" will only support a single "output". This change is only relevant to how the Agent abstract the multiples outputs cases.

**Pull request merged**